### PR TITLE
Upgrade ParamParser to make it more convinient.

### PR DESCRIPTION
**Pull request by f**
It now decouples `env.params` to `env.params.query`, `env.params.body`,
`env.params.json` and `env.params.url` and you still can access merged
values using `env.params.all`

### CODE REVIEW

File: src/kemal/param_parser.cr [hunk 1]
```diff
@@ -5,16 +5,35 @@ require "json"
...
+
+  def initialize
+    @url = {} of String => String
+    @query = {} of String => String
+    @body = {} of String => String
+    @json = {} of String => AllParamTypes
+  end
+
+  def all
+    @url.merge(@query).merge(@body).merge(@json)
```
Comment (jmoriau) [inline src/kemal/param_parser.cr]: I would have merged the `@url` last so that it doesn't get overridden by an unwanted parameter in the body.

**Comment by sdogruyol** [comment]
I wonder if we can improve this.

Currently it has %20 lower performance compared to current `ParamParser`.

```ruby
require "benchmark"
require "./kemal/*"

request = HTTP::Request.new(
  "POST",
  "/?foo=bar",
  body: "[1]",
  headers: HTTP::Headers{"Content-Type": "application/json"},
)

Benchmark.ips do |x|
  x.report "Old one" do
    params = Kemal::ParamParser.new(request).parse
  end
end
```

Old one 214.12k IPS
New one 174.87k IPS

File: src/kemal/param_parser.cr (hunk 1 above)
Comment (sdogruyol) [inline src/kemal/param_parser.cr]: That makes sense.

Comment (f) [inline src/kemal/param_parser.cr]: I made it in the order of severity. But when I think again URL is much more important. Makes sense.

Comment (jmoriau) [inline src/kemal/param_parser.cr]: :stuck_out_tongue_winking_eye:

**Comment by sdogruyol** [comment]
This is awesome! Compared to current `ParamParser`
- At least %40 faster in worst scenario 
- At most %400 fast in best scenario

Thanks @f :+1:

**Comment by f** [comment]
@sdogruyol Wooha! We made a big change! It's blazing fast :rocket:

**Comment by jmoriau** [comment]
It's interesting though, the changes are relatively small and 40% is a big improvement, did you pinpoint where/why in the worst case it's still faster even though it's still parsing everything and it's even allocating 3 more Hash ?

**Comment by f** [comment]
@jmoriau Actually it's all about memoizing.

**Comment by jmoriau** [comment]
You mean because you only parse once and store the value ? It's weird because even though it's a lot more readable and a bit easier to use, it looks like it's doing the same thing as before (in the worst case scenario where you would make it parse everything). Im really curious I'll run some test tomorrow :smile: but good job :+1: looks a lot better!

**Comment by f** [comment]
@jmoriau It wasn't lazily parsing actually. For example, if you want to get "`x`" parameter of URL (`/abc/:x`) it will also try to parse `body` and `json` and others. It was lazy but not **this** clever. Since we decoupled `body`, `url`, `json` and `body`, they are all lazy but not together.

In worst case scenario it's faster because of **memoizing**. The parsed parameters memoized once and never parse again. Before there was no memoization.

**Comment by sdogruyol** [comment]
LGTM :+1: :rocket:

**Pull request merged**